### PR TITLE
add ViTResiDual

### PR DESCRIPTION
**Pull request by Hazqeel09**
I implemented ResiDual Connection from this <a href="https://arxiv.org/abs/2304.14802">paper</a>. It added pre and post layer normalization to transformer architecture.

### CODE REVIEW

File: vit_pytorch/vit_residual.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                Attention(dim, heads = heads, dim_head = dim_head, dropout = dropout),
+                FeedForward(dim, mlp_dim, dropout = dropout)
+            ]))
+    def forward(self, x):
+        for attn, ff in self.layers:
+            res = x * self.enc_res_input_norm_scale
+            xr = x
+
+            x = attn(x)
+            res = res + x * self.enc_res_input_norm_scale
```
Comment (lucidrains) [inline vit_pytorch/vit_residual.py:77]: it seems like you never use the `res` again

**Comment by lucidrains** [comment]
Screen Shot 2023-05-09 at 8 57 45 AM

it trains at the very least, not much discernable difference

**Comment by lucidrains** [comment]
@Hazqeel09 i've added it [here](https://github.com/lucidrains/x-transformers#residual) if you want to play around with it / run some experiments

**Comment by lucidrains** [comment]
@Hazqeel09 have you run any experiments yet? does the paper check out?

**Comment by Hazqeel09** [comment]
@lucidrains not yet, I'm currently running different experiments,  will update if I do

**Comment by lucidrains** [comment]
@Hazqeel09 ok! do send me an email if you see something interesting

would be also nice to compare it to the deep vit paper, where they also approached the representational collapse issue from another angle

**Comment by Hazqeel09** [comment]
@lucidrains I have tried residual for my research (understand embedding context, where I tried to replace CNN with ViT from this  [paper](https://ieeexplore.ieee.org/document/9658729) but it didnt help to produce good result so I tried with [spectformer](https://arxiv.org/abs/2304.06446) where it use spectral block to understand local context and attention to understand global context.

ViT and GFNet works better for my research but not spectformer, so I decided to compare them using image dataset. You can see the result [here](https://github.com/Hazqeel09/VisionModelsBenchmark/blob/main/Benchmarking.ipynb) and the code for vanilla spectformer architecture [here](https://github.com/Hazqeel09/VisionModelsBenchmark/blob/main/vit_flash_spectformer_real.py).

If you consider the code is good, I will modify the code a bit to remove flash attention and then submit new PR.

**Pull request closed without merging**